### PR TITLE
DataStores should return the UID of the data entry back to the user

### PR DESCRIPTION
**Pull request by awharrison-28**
### Motivation and Context
As we continually add more storage connectors to the Semantic Kernel, we need to be prepared for how they behave. In particular, vector storage service support querying data by a unique database ID and by metadata match. This PR separates the id of a memory ID (searchable metadata) from a DataEntry key. **This unique DataEntry key is to be generated at the IMemoryStore() implementor level**. Some databases will generate this for the user, while others will expect it to be generated by the user and passed to the DB during an upsert request. 

Out of scope (coming soon!): 

- Adding methods to support metadata search
- Ensuring vector uniqueness based on metadata id

### Description

- PutAsync no longer returns the full object placed in the datastore. It returns the unique database ID of the DataEntry. All methods that rely on PutAsync bubble up this UID to the user. 
- Modified PutAsync to behave as upsert and added parameters to callers of PutAsync to optionally provided a DataEntry key
- Modified Datastore unit tests to reflect the new behavior

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Skills/Skills.Memory.Sqlite/SqliteMemoryStore.cs [hunk 1]
```diff
@@ -66,10 +66,21 @@ await foreach (DatabaseEntry dbEntry in this._dbConnection.ReadAllAsync(collecti
...
-        return data;
+        string key;
+        if (string.IsNullOrEmpty(data.Key))
+        {
+            key = Guid.NewGuid().ToString();
+        }
+        else
+        {
+            key = data.Key;
+            await this._dbConnection.DeleteAsync(collection, key, cancel);
```
Comment (adrianwyatt) [inline dotnet/src/SemanticKernel.Skills/Skills.Memory.Sqlite/SqliteMemoryStore.cs:79]: What is the opportunity for a key collision here? In the case where there is a collision, the delete succeeds and the insert fails, we lose data yes? If this is the rare scenario, we may want to at least log that it happened.

File: dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs [hunk 2]
```diff
@@ -146,7 +146,7 @@ public string Recall(string text, SKContext context)
     [SKFunctionInput(Description = "The information to save")]
     [SKFunctionContextParameter(Name = CollectionParam, Description = "Memories collection associated with the information to save", DefaultValue = DefaultCollection)]
     [SKFunctionContextParameter(Name = KeyParam, Description = "The key associated with the information to save")]
-    public async Task SaveAsync(string text, SKContext context)
+    public async Task<string> SaveAsync(string text, SKContext context)
```
Comment (adrianwyatt) [inline dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs:149]: Could we add an SKFunctionDescription that describes the return value?

File: dotnet/src/SemanticKernel/Memory/ISemanticTextMemory.cs [hunk 3]
```diff
@@ -15,32 +15,36 @@ public interface ISemanticTextMemory
...
     /// </summary>
     /// <param name="collection">Collection where to save the information</param>
-    /// <param name="id">Unique identifier</param>
+    /// <param name="id">Unique identifier for searching memory metadata</param>
     /// <param name="text">Information to save</param>
     /// <param name="description">Optional description</param>
+    /// <param name="dbKey">The unique database key.</param>
     /// <param name="cancel">Cancellation token</param>
-    public Task SaveInformationAsync(
+    public Task<string> SaveInformationAsync(
```
Comment (adrianwyatt) [inline dotnet/src/SemanticKernel/Memory/ISemanticTextMemory.cs:23]: Could we add <returns> tags for methods in this interface that now return a value?

File: dotnet/src/SemanticKernel/Memory/Storage/DataEntry.cs [hunk 4]
```diff
@@ -33,7 +33,7 @@ public DataEntry(string key, TValue? value, DateTimeOffset? timestamp = null)
     /// Gets the key of the data.
     /// </summary>
     [JsonPropertyName("key")]
-    public readonly string Key { get; }
+    public string? Key { get; set; } = null;
```
Comment (adrianwyatt) [inline dotnet/src/SemanticKernel/Memory/Storage/DataEntry.cs:36]: Does it make sense to allow the Key to be publicly set when we expect it to be set in the constructor? I'm curious on the scenario where we would modify the model after construction.

**Review by adrianwyatt — CHANGES_REQUESTED** [review]

File: dotnet/src/SemanticKernel/Memory/Storage/DataEntry.cs [hunk 5]
```diff
@@ -202,7 +206,7 @@ public static DataEntry<TValue> Create<TValue>(string key, TValue? value, DateTi
     /// <param name="value">The data value.</param>
     /// <param name="timestamp">The data timestamp.</param>
     /// <returns>A <see cref="DataEntry{TValue}"/> object.</returns>
-    public static DataEntry<TValue> Create<TValue>(string key, string value, DateTimeOffset? timestamp = null)
+    public static DataEntry<TValue> Create<TValue>(string? key, string value, DateTimeOffset? timestamp = null)
     {
         Verify.NotEmpty(key, "Data entry key cannot be NULL");
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel/Memory/Storage/DataEntry.cs:211]: Remove like above?

**Comment by awharrison-28** [comment]
Pausing PR due to incoming memory interfaces refactor.

**Review by adrianwyatt — APPROVED** [review]

File: dotnet/src/SemanticKernel/Memory/SemanticTextMemory.cs [hunk 6]
```diff
@@ -28,32 +28,34 @@ public sealed class SemanticTextMemory : ISemanticTextMemory, IDisposable
...
 
     /// <inheritdoc/>
-    public async Task SaveReferenceAsync(
+    public async Task<string> SaveReferenceAsync(
         string collection,
         string text,
         string externalId,
         string externalSourceName,
         string? description = null,
+        string? dbKey = null,
```
Comment (dluc) [inline dotnet/src/SemanticKernel/Memory/SemanticTextMemory.cs:52]: why not use externalId? wouldn't it solve the duplication problem?

**Review by dluc — CHANGES_REQUESTED** [review]
the memory duplication problem seems a big issue, what are we doing about it? also why do we need another unique key when calling SaveInformationAsync and SaveReferenceAsync?

**Comment by awharrison-28** [comment]
closing in favor of https://github.com/microsoft/semantic-kernel/pull/308

**Pull request closed without merging**